### PR TITLE
[MAID-3158] feat/nonce: allow encryption using custom nonces

### PR DESCRIPTION
**Pull request by mrcnski**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -581,9 +608,43 @@ impl Default for SymmetricKey {
...
+    }
+
+    /// Create a nonce from bytes. Useful when it has been serialised.
+    pub fn from_bytes(nonce: [u8; NONCE_BYTES]) -> Self {
+        Self {
+            nonce: secretbox::Nonce(nonce),
+        }
+    }
+
+    /// Convert the `SymmetricNonce` into the raw underlying bytes.
```
Comment (nbaksalyar) [inline src/lib.rs]: `SymmetricNonce` - probably supposed to be just `Nonce`?

File: src/lib.rs [hunk 2]
```diff
@@ -111,6 +111,10 @@ pub const SHARED_SECRET_KEY_BYTES: usize = box_::PRECOMPUTEDKEYBYTES;
 pub const SIGNATURE_BYTES: usize = sign::SIGNATUREBYTES;
 /// Symmetric key length in bytes.
 pub const SYMMETRIC_KEY_BYTES: usize = secretbox::KEYBYTES;
+/// Nonce length in bytes.
+pub const NONCE_BYTES: usize = box_::NONCEBYTES;
```
Comment (nbaksalyar) [inline src/lib.rs]: Shouldn't it be `secretbox::NONCEBYTES`?

File: src/lib.rs [hunk 3]
```diff
@@ -581,9 +608,43 @@ impl Default for SymmetricKey {
     }
 }
 
+/// Nonce structure used for authenticated symmetric encryption.
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone)]
+pub struct Nonce {
+    nonce: secretbox::Nonce,
```
Comment (nbaksalyar) [inline src/lib.rs:614]: A thing to note here: if this nonce is to be used for asymmetric encryption, we'll need to be sure that it's compatible with `rust_sodium::box_::Nonce`.

**Review by nbaksalyar — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 3 above)
Comment (mrcnski) [reply src/lib.rs:614]: I don't think we want a nonce for asymmetric encryption. Nonces are only used by symmetric encryption -- correct me if I'm wrong. I initially had a separate nonce for this case, hence the other things you pointed out, but I removed it.

Edit: I see nonces are used for precomputing a shared secret key. Should we support custom nonces for this case as well?

**Review by nbaksalyar — DISMISSED** [review]

**Comment by mrcnski** [comment]
Closing, we merged these commits in #25

**Pull request closed without merging**